include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -1277,6 +1318,128 @@ TensorProductMatrixSymmetricSumCollection<dim, Number, n_rows_1d>::finalize()
 
       cache.clear();
     }
+
+  if (precomute_inverse_diagonal)
+    {
+      if (dim == 1)
+        {
+          this->inverted_eigenvalues.resize(this->eigenvalues.size());
+          for (unsigned int i = 0; i < this->eigenvalues.size(); ++i)
+            this->inverted_eigenvalues[i] = Number(1.0) / this->eigenvalues[i];
+        }
+      else
+        {
+          std::vector<unsigned int> indices_ev;
+
+          if (indices.size() > 0)
+            {
+              const unsigned int n_cells = indices.size() / dim;
+
+              // create cache for eigenvalues and store index of each cell
+              std::map<std::array<unsigned int, dim>, unsigned int> cache_ev;
+              std::vector<unsigned int> cache_ev_idx(n_cells);
```

Comment:
Without a comment, it is unclear why there is a need for an additional search and what the code here is doing in addition to the one above. I believe I understand what is going on here (we need to make sure that all `dim` matrices really point to the same place as we build the product), but it took me quite some time. Maybe this is something that could be done in a separate named function to improve the code readability?

Reply:
I have added some comments. I hope this is better now.